edition = "2023";

message GameMessage {
	oneof msg {
		ServerInstruction serverInstruction = 1;
		CharacterDiscovery characterDiscovery = 2;
		CharacterVannish characterVannish = 3;
		CharacterMove characterMove = 4;
		CharacterAction characterAction = 5;
		CharacterVitals characterVitals = 6;
		CharacterStats characterStats = 7;
		CharacterAppeal characterAppeal = 8;
	}
}

message ServerInstruction {
	uint32 instruction = 1;
}

message CharacterDiscovery {
	uint32 id = 1;
	string name = 2;
	string char_uid = 3;
	uint32 char_type = 4;
	uint32 char_specific = 5;
	uint32 loaction_area = 6;
}

message CharacterDiscoverySeveral {
	repeated CharacterDiscovery characters = 1;
}

message CharacterVannish {
	repeated uint32 id = 1;
}

message CharacterMove {
	uint32 id = 1;
	float x = 2;
	float y = 3;
	float z = 4;
	float mx = 5;
	float my = 6;
	float mz = 7;
	float facing = 8;
	uint32 move_type = 9;
	uint32 loaction_area = 10;
}

message CharacterAction {
	uint32 id = 1;
}

message CharacterVitals {
	uint32 id = 1;
	uint32 health = 2;
	uint32 power = 3;
	map<uint32, uint32> debuffs = 4;
	map<uint32, uint32> buffs = 5;
}

message CharacterStats {
	uint32 id = 1;
	uint32 health = 2;
	uint32 power = 3;
	uint32 max_health = 4;
	uint32 max_power = 5;
	uint32 phys_damage = 6;
	uint32 tact_damage = 7;
	uint32 phys_defence = 8;
	uint32 tact_defence = 9;
}

message CharacterAppeal {
	uint32 id = 1;
}
